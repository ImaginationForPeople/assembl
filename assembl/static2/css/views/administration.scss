.administration {
  margin: $margin-section + $navbar-height 0;

  .admin-menu-container {
    height: 350px;
    position: relative;
  }

  .admin-menu {
    padding: 0;
    position: fixed;
    list-style-type: none;

    li {
      padding: 0 0 15px;

      a {
        color: $grey2;

        &:hover {
          color: $first-color;
        }
      }
    }

    .active {
      color: $first-color;
      text-decoration: none;
    }
  }

  .admin-menu2 {
    list-style-type: none;
    padding: 15px 0 0 30px;

    li {
      &:last-child {
        padding: 0;
      }
    }
  }

  .language-menu {
    position: fixed;
    background-color: $white;
    padding-bottom: 15px;
    margin-top: 20px;
  }

  .admin-help {
    font-size: $font-size-xs;
    font-family: $font-family6;
    margin-top: -15px;
    letter-spacing: 0;
  }

  .admin-box {
    background-color: $white;

    .dark-title-3 {
      padding: 40px 40px 0;
    }

    .admin-content {
      padding: 30px 40px 80px;
      text-align: center;
    }

    .admin-language-content {
      padding: 30px 30% 80px;
      text-align: left;

      .language-list {
        padding-top: 20px;
        color: $grey2;
      }
    }

    .two-columns-admin {
      display: flex;
      margin: 40px 0 0;

      .column-left {
        flex: 1;
      }

      .column-right {
        flex: 1;
      }
    }

    .annotation {
      margin: 0 40px 20px;
    }
  }

  .admin-icons {
    border: none;
    width: 35px;

    &:hover {
      background-color: $white;
      border: none;
    }

    &:focus {
      outline: none;
      background-color: $white;
    }

    &:active {
      box-shadow: none;
    }
  }

  .plus {
    background-color: $first-color;
    border-radius: 20px;
    color: $white;
    cursor: pointer;
    height: 20px;
    margin-left: auto;
    margin-right: auto;
    padding: 0 0 1px 2px;
    text-align: center;
    width: 20px;
  }

  .form-container {
    margin: 0 auto;
    max-width: 400px;
    text-align: left;

    .title {
      font-size: 16px;
      margin: 0 0 25px;
      text-align: left;
    }

    .form-control {
      font-size: 12px;
    }
  }

  .flag-container {
    cursor: pointer;
    margin: 20px 15px 0;

    &:hover {
      transform: scale(1.2);
    }
  }

  .flag-container.active {
    border-width: $btn-dark-border-width;
    border-style: $btn-dark-border-style;
    border-color: $btn-dark-border-color;
    border-radius: 20px;
    height: 24px;
    margin: 20px 15px 0 13px;
    transform: scale(1.2);
    width: 24px;
  }

  .assembl-icon-delete {
    &:before {
      font-size: 15px;
    }
  }

  .separator {
    border-bottom: 1px $btn-dark-border-style $btn-dark-border-color;
    clear: both;
    margin: 0 0 50px;
    padding: 40px 0 15px;
  }

  .admin-navbar {
    background-color: $first-color-light;
    height: $navbar-height;
    width: 100%;

    .step-numbers {
      max-width: 250px;
      padding: 0 0 0 40px;
      position: relative;
      width: 100%;

      .txt {
        color: $first-color;
        font-family: $font-family4;
        font-size: $font-size-xs;
        margin: 12px 0 5px;
      }

      .bar {
        background-color: $first-color;
        height: 6px;
        position: absolute;
        z-index: 2;
      }

      .bkg-bar {
        background-color: $white;
        height: 6px;
        width: 100%;
        position: absolute;
        z-index: 1;
      }
    }
  }

  .arrow-container {
    padding: 0 40px 0 0;

    .arrow {
      background-color: $white;
      border-radius: 26px;
      cursor: pointer;
      display: inline-block;
      line-height: 20px;
      height: 26px;
      margin: 18px 0 0 30px;
      padding: 0 0 0 4px;
      text-align: left;
      transform: rotate(90deg);
      width: 26px;

      .assembl-icon-up-open {
        color: $first-color;

        &:before {
          font-size: 8px;
        }
      }

      .assembl-icon-down-open {
        color: $first-color;

        &:before {
          font-size: 8px;
        }
      }
    }
  }

  .save-bar {
    border-bottom: $navbar-border;
    border-top: $navbar-border;
    background-color: $white;
    height: $navbar-height;
    padding: 10px 0;
    position: fixed;
    top: $navbar-height;
    width: 100%;
    z-index: 100;
  }

  .save-button {
    position: fixed;
    margin-top: -102px;
    right: 0;
    z-index: 101;
  }

  .tab-title {
    border-bottom: $border-not-active;
    color: $grey2;
    cursor: pointer;
    display: block;
    padding: 0 0 5px 0;
  }

  .tab-title-active {
    border-bottom: $border-active;
    color: $grey3;
    cursor: pointer;
    display: block;
    padding: 0 0 5px 0;
    text-decoration: none;
  }

  .text-area {
    height: 100px;
  }

  .preview {
    border-radius: 5px;
    margin: 25px 0 0;

    img {
      border-radius: 5px;
      max-width: 390px;
    }
  }

  .preview-title {
    color: $grey3;
    font-family: $font-family6;
    font-size: $font-size-xs;
    margin: 10px 0 20px 5px;
  }

  .question-section {
    margin: 0 0 75px 0;
  }

  label {
    font-family: $font-family7;
    font-size: $font-size-xs;
    margin: 0 0 5px 5px;
  }

  .help-block {
    font-size: $font-size-xs;
    margin: 5px 0 0 5px;
  }

  .assembl-icon-faq {
    &:before {
      font-size: 12px;
    }

    &:hover {
      color: $grey3;
    }
  }

  .token-vote-form {
    .flex > .form-group {
      flex-grow: 1;
    }
  }

  .landing-page-admin {
    .admin-box {
      .admin-content {
        text-align: left;
      }
    }

    .form-container {
      max-width: 700px;
    }

    .modules-preview {
      margin: 5px;
      padding: 5px;
      background-color: $grey0;
      min-height: 400px;
      display: flex;
      flex-direction: column;
      justify-content: center;

      .inner {
        margin: auto;
        border: 1px solid black;
        width: 70%;
        min-height: 330px;
        display: flex;
        flex-direction: column;
        justify-content: space-between;

        .module-block:first-child {
          padding: 10px 0;
        }

        .module-block:first-child,
        .module-block:last-child {
          background-color: $first-color-light;
          color: $white;
          border: 0;
          margin: 0;
          text-align: center;
        }

        .other-modules {
          flex: 1;
          justify-content: flex-start;

          .module-block {
            margin: 7px;
            padding: 15px 10px 15px 16px;
            border: 0.8px solid $first-color;
            color: $first-color;
            background-color: $white;
            display: flex;
            justify-content: space-between;

            .admin-icons {
              color: $first-color;
              padding: 0;
              width: 20px;
            }
          }
        }
      }
    }
  }
}

.helper-popover {
  width: auto;
  min-width: 332px;
  padding-top: 8px;

  .arrow {
    display: none;
  }

  img {
    padding-bottom: 10px;
  }
}
