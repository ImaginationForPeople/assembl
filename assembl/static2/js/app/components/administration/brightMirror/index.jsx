// @flow
import React from 'react';
import { connect } from 'react-redux';
import { I18n } from 'react-redux-i18n';
import arrayMutators from 'final-form-arrays';
import { type ApolloClient, compose, withApollo } from 'react-apollo';
import { Field } from 'react-final-form';

// Components
import FileUploaderFieldAdapter from '../../form/fileUploaderFieldAdapter';
import MultilingualTextFieldAdapter from '../../form/multilingualTextFieldAdapter';
import MultilingualRichTextFieldAdapter from '../../form/multilingualRichTextFieldAdapter';
import LoadSaveReinitializeForm from '../../../components/form/LoadSaveReinitializeForm';
import SaveButton from '../../../components/administration/saveButton';
import Loader from '../../common/loader';
import Helper from '../../common/helper';

// Functions
import { load, postLoadFormat } from './load'; // Load file needs to be updated according to bright mirror requirements
import { createMutationsPromises, save } from './save'; // Save file needs to be updated according to bright mirror requirements
import validate from './validate'; // Save file needs to be updated according to bright mirror requirements

type Props = {
  client: ApolloClient,
  editLocale: string
};

const loading = <Loader />;

class BrightMirrorAdminForm extends React.Component<Props> {
  render() {
    const { client, editLocale } = this.props;
    const name = 'bright-mirror';
    return (
      <LoadSaveReinitializeForm
        load={(fetchPolicy: FetchPolicy) => load(client, fetchPolicy)}
        loading={loading}
        postLoadFormat={postLoadFormat}
        createMutationsPromises={createMutationsPromises(client)}
        save={save}
        validate={validate}
        mutators={{
          ...arrayMutators
        }}
        render={({ handleSubmit, pristine, submitting }) => (
          <div className="admin-content">
            <form onSubmit={handleSubmit}>
              <SaveButton disabled={pristine || submitting} saveAction={handleSubmit} />
              <div className="form-container">
                <Helper
                  label={I18n.t('administration.headerTitle')}
                  helperUrl="/static2/img/helpers/helper2.jpg"
                  helperText="TO BE DEFINED and IMAGE TO BE CHANGED (translation to add in translation.js)"
                  classname="title"
                />
                <Field
                  editLocale={editLocale}
                  name={`${name}.title`}
                  component={MultilingualTextFieldAdapter}
                  label={`${I18n.t('administration.brightMirrorSection.thematicTitleLabel')} ${editLocale.toUpperCase()}`}
                  required
                />
                <Field
                  editLocale={editLocale}
                  name={`${name}.title`}
                  component={MultilingualTextFieldAdapter}
                  label={`${I18n.t('administration.brightMirrorSection.bannerSubtitleLabel')} ${editLocale.toUpperCase()}`}
                  required
                />
                <Field
                  name={`${name}.img`}
                  component={FileUploaderFieldAdapter}
                  label={I18n.t('administration.brightMirrorSection.bannerImagePickerLabel')}
                />
                <Helper
                  label={I18n.t('administration.instructions')}
                  helperUrl="/static2/img/helpers/helper2.jpg"
                  helperText="TO BE DEFINED and IMAGE TO BE CHANGED (translation to add in translation.js)"
                  classname="title"
                />
                <Field
                  editLocale={editLocale}
                  name={`${name}.title`}
                  component={MultilingualTextFieldAdapter}
                  label={`${I18n.t('administration.brightMirrorSection.sectionTitleLabel')} ${editLocale.toUpperCase()}`}
                  required
                />
                <Field
                  key={`${name}-text-${editLocale}`}
                  editLocale={editLocale}
                  name={`${name}.text`}
                  component={MultilingualRichTextFieldAdapter}
                  label={`${I18n.t('administration.brightMirrorSection.instructionLabel')} ${editLocale.toUpperCase()}`}
                />
              </div>
            </form>
          </div>
        )}
      />
    );
  }
}

const mapStateToProps = ({ admin: { editLocale, resourcesCenter } }) => {
  const { page, resourcesById, resourcesHaveChanged, resourcesInOrder } = resourcesCenter;
  return {
    editLocale: editLocale,
    pageHasChanged: page.get('_hasChanged'),
    resourcesCenterPage: page,
    resourcesHaveChanged: resourcesHaveChanged,
    resources: resourcesInOrder.map(id => resourcesById.get(id).toJS())
  };
};

export default compose(connect(mapStateToProps), withApollo)(BrightMirrorAdminForm);