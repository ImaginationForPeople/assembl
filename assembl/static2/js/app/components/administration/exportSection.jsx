// @flow
import * as React from 'react';
import { I18n, Translate } from 'react-redux-i18n';
import moment from 'moment';
import classnames from 'classnames';
import { Link } from 'react-router';
import { FormGroup, Radio, Checkbox, FormControl } from 'react-bootstrap';

import SectionTitle from './sectionTitle';
import CustomDateRangePicker from './dateRangePicker/customDateRangePicker';
import { datePickerPresets } from '../../constants';
import { getFullDebatePreset } from '../form/utils';

type Props = {
  exportLink: string,
  annotation: string,
  sectionTitle: string,
  locale?: ?string,
  exportLocale?: ?string,
  languages?: Array<Language>,
  handleDatesChange?: ?Function,
  handleAnonymousChange?: ?Function,
  handleExportLocaleChange?: ?Function,
  disableExportButton: Function,
  buttonIsDisabled: boolean,
  shouldBeAnonymous?: boolean,
  handleShouldTranslate?: ?Function,
  shouldTranslate?: ?boolean,
  start?: ?moment$Moment,
  end?: ?moment$Moment,
  phases?: ?Timeline
};

export const ExportSection = ({
  annotation,
  sectionTitle,
  exportLink,
  exportLocale,
  handleDatesChange,
  handleExportLocaleChange,
  handleShouldTranslate,
  handleAnonymousChange,
  shouldBeAnonymous,
  shouldTranslate,
  start,
  end,
  phases,
  languages,
  locale,
  disableExportButton,
  buttonIsDisabled
}: Props) => {
  const renderAnonymousOption = () =>
    (handleAnonymousChange ? (
      <React.Fragment>
        <Translate value="administration.export.anonymity" />
        <Checkbox onChange={handleAnonymousChange} value={shouldBeAnonymous}>
          <Translate value="administration.export.anonymous" />
        </Checkbox>
      </React.Fragment>
    ) : null);

  const renderLanguageOptions = () => {
    if (!handleExportLocaleChange || !languages || !handleShouldTranslate) {
      return null;
    }
    const activeLanguage = languages ? languages.filter(language => language.locale === exportLocale)[0] : null;
    return (
      <React.Fragment>
        <Translate value="administration.export.translation" />
        <Radio checked={!shouldTranslate} onChange={() => handleShouldTranslate(false)}>
          <Translate value="administration.export.noExportLanguage" />
        </Radio>
        <Radio checked={shouldTranslate} onChange={() => handleShouldTranslate(true)}>
          <Translate value="administration.export.translateTheMessagesIn" />
          {shouldTranslate && (
            <FormControl
              className="export-language-dropdown"
              componentClass="select"
              onChange={(e) => {
                handleExportLocaleChange(e.target.value);
              }}
              value={activeLanguage ? activeLanguage.locale : ''}
            >
              {languages &&
                languages.map(lang => (
                  <option key={`locale-${lang.locale}`} value={lang.locale}>
                    {lang.name}
                  </option>
                ))}
            </FormControl>
          )}
        </Radio>
      </React.Fragment>
    );
  };

  const renderDatePicker = () => {
    if (!locale || !handleDatesChange) {
      return null;
    }
    const phasesPresets = phases
      ? phases.map((phase, index) => ({
        id: index + 1,
        labelTranslationKey: 'administration.export.presets.phase',
        range: {
          startDate: moment(phase.start),
          endDate: moment(phase.end)
        },
        type: 'phase'
      }))
      : [];
    const fullDebatePreset = phasesPresets && phasesPresets.length > 0 && getFullDebatePreset(phasesPresets);
    const presets = fullDebatePreset ? [...datePickerPresets, ...phasesPresets, fullDebatePreset] : [...datePickerPresets];
    return presets ? (
      <div className="export-date">
        <Translate value="administration.export.exportDate" />
        <CustomDateRangePicker presets={presets} locale={locale} handleDatesChange={handleDatesChange} start={start} end={end} />
      </div>
    ) : null;
  };

  const exportButtonClassNames = classnames('button-link button-dark margin-l', { disabled: buttonIsDisabled });

  return (
    <div className="admin-box admin-export-section">
      <SectionTitle title={I18n.t(`administration.export.${sectionTitle}`)} annotation={I18n.t(`administration.export.${annotation}`)} />
      <div className="admin-content">
        <FormGroup>
          <div className="export-options">
            <div>
              {renderAnonymousOption()}
              {renderLanguageOptions()}
            </div>
            {renderDatePicker()}
          </div>
        </FormGroup>
        <div className="center-flex">
          <Link className={exportButtonClassNames} href={exportLink} onClick={disableExportButton}>
            <Translate value="administration.export.link" />
          </Link>
        </div>
      </div>
    </div>
  );
};

ExportSection.defaultProps = {
  annotation: 'defaultAnnotation',
  sectionTitle: 'defaultSectionTitle',
  locale: null,
  exportLocale: null,
  languages: [],
  handleDatesChange: null,
  handleAnonymousChange: null,
  handleExportLocaleChange: null,
  handleShouldTranslate: null,
  shouldBeAnonymous: false,
  shouldTranslate: false,
  start: null,
  end: null,
  phases: null
};

export default ExportSection;