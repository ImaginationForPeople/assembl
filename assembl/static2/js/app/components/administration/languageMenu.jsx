import React from 'react';
import { connect } from 'react-redux';
import { OverlayTrigger } from 'react-bootstrap';

import { languageTooltip } from '../common/tooltips';
import { updateEditLocale } from '../../actions/adminActions';
import En from '../svg/flags/en';
import Fr from '../svg/flags/fr';
import Ja from '../svg/flags/ja';
import ZhCN from '../svg/flags/zh_CN';
import Ru from '../svg/flags/ru';

// TO DO get it dynamically
const Flag = ({ locale }) => {
  switch (locale) {
  case 'en':
    return <En />;
  case 'fr':
    return <Fr />;
  case 'ja':
    return <Ja />;
  case 'ru':
    return <Ru />;
  case 'zh_Hans':
    return <ZhCN />;
  default:
    return <span>{locale}</span>;
  }
};

const LanguageMenu = ({ changeLocale, editLocale, discussionPreferences, isHidden }) => {
  if (!isHidden) {
    return (
      <div className="relative">
        <div className="language-menu">
          <OverlayTrigger placement="top" overlay={languageTooltip}>
            <div>
              {discussionPreferences.map((key, index) => (
                <div
                  onClick={() => changeLocale(key)}
                  id={key}
                  className={editLocale === key ? 'flag-container active' : 'flag-container'}
                  key={index}
                >
                  <Flag locale={key} />
                </div>
              ))}
            </div>
          </OverlayTrigger>
        </div>
      </div>
    );
  }

  return <span />;
};

const mapStateToProps = state => ({
  translations: state.i18n.translations,
  editLocale: state.admin.editLocale,
  discussionPreferences: state.admin.discussionLanguagePreferences,
  isHidden: state.admin.displayLanguageMenu
});

const mapDispatchToProps = dispatch => ({
  changeLocale: (newLocale) => {
    dispatch(updateEditLocale(newLocale));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(LanguageMenu);