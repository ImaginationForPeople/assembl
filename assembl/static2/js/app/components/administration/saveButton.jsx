import React from 'react';
import { connect } from 'react-redux';
import { graphql, withApollo, compose } from 'react-apollo';
import { Button } from 'react-bootstrap';
import { Translate, I18n } from 'react-redux-i18n';

import { displayAlert } from '../../utils/utilityManager';
import { createThematic, updateThematic, deleteThematic } from '../../graphql/ThematicMutations';

const runSerial = (tasks) => {
  let result = Promise.resolve();
  tasks.forEach((task) => {
    result = result.then(task);
  });
  return result;
};

const createVariablesForMutation = (thematic) => {
  // If imgUrl is an object, it means it's a File.
  // We need to send image: null if we didn't change the image.
  return {
    identifier: 'survey',
    titleEntries: thematic.titleEntries,
    image: typeof thematic.imgUrl === 'object' ? thematic.imgUrl : null,
    video: thematic.video,
    questions: thematic.questions
  };
};

const SaveButton = ({ createThematic, deleteThematic, enabled, refetchThematics, thematics, updateThematic }) => {
  const saveAction = () => {
    const promisesArray = [];
    thematics.forEach((thematic) => {
      if (thematic.isNew && !thematic.toDelete) {
        // create thematic
        const payload = {
          variables: createVariablesForMutation(thematic)
        };
        const p1 = () => {
          return createThematic(payload);
        };
        promisesArray.push(p1);
      } else if (thematic.toDelete && !thematic.isNew) {
        // delete thematic
        const payload = {
          variables: {
            thematicId: thematic.id
          }
        };
        const p3 = () => {
          return deleteThematic(payload);
        };
        promisesArray.push(p3);
      } else {
        // update thematic
        const variables = createVariablesForMutation(thematic);
        variables.id = thematic.id;
        const payload = {
          variables: variables
        };
        const p2 = () => {
          return updateThematic(payload);
        };
        promisesArray.push(p2);
      }
    });

    runSerial(promisesArray)
      .then(() => {
        refetchThematics();
        displayAlert('success', I18n.t('administration.successThemeCreation'));
      })
      .catch((error) => {
        displayAlert('danger', `${error}`);
      });
  };

  return (
    <Button className="button-submit button-dark right" disabled={!enabled} onClick={saveAction}>
      <Translate value="administration.saveThemes" />
    </Button>
  );
};

const SaveButtonWithMutations = compose(
  graphql(createThematic, {
    name: 'createThematic'
  }),
  graphql(updateThematic, {
    name: 'updateThematic'
  }),
  graphql(deleteThematic, {
    name: 'deleteThematic'
  })
)(SaveButton);

const mapStateToProps = ({ admin: { thematicsById, thematicsHaveChanged, thematicsInOrder } }) => {
  return {
    enabled: thematicsHaveChanged,
    thematics: thematicsInOrder.toArray().map((id) => {
      return thematicsById.get(id).toJS();
    })
  };
};

export default compose(connect(mapStateToProps), withApollo)(SaveButtonWithMutations);