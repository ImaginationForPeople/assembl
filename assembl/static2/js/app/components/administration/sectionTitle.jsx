// @flow
import React from 'react';

type SectionTitleProps = {
  title: string,
  annotation: string
};

const SectionTitle = ({ title, annotation }: SectionTitleProps) => (
  <div>
    <h3 className="dark-title-3">{title}</h3>
    <div className="box-hyphen" />
    <div className="annotation">{annotation}</div>
  </div>
);

export default SectionTitle;