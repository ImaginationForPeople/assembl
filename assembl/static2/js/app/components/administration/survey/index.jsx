// @flow
import React from 'react';
import isEqualWith from 'lodash/isEqualWith';
import { type ApolloClient, compose, withApollo } from 'react-apollo';
import { connect } from 'react-redux';
import arrayMutators from 'final-form-arrays';

import AdminForm from '../../form/adminForm';
import { compareEditorState } from '../../form/utils';
import swap from '../../form/swap';
import LoadSaveReinitializeForm from '../../form/LoadSaveReinitializeForm';
import Step1 from './step1';
import ConfigureThematicForm from './configureThematicForm';
import { load, postLoadFormat } from './load';
import { createMutationsPromises, save } from './save';
import validate from './validate';
import Loader from '../../common/loader';

type Props = {
  client: ApolloClient,
  phaseIdentifier: string,
  section: string,
  thematicId: string,
  discussionPhaseId: string,
  editLocale: string,
  locale: string
};

const loading = <Loader />;

const DumbSurveyAdminForm = ({ client, phaseIdentifier, section, thematicId, discussionPhaseId, editLocale, locale }: Props) => {
  if (!discussionPhaseId) {
    return loading;
  }
  return (
    <LoadSaveReinitializeForm
      load={(fetchPolicy: FetchPolicy) => load(client, fetchPolicy, discussionPhaseId, locale)}
      loading={loading}
      postLoadFormat={postLoadFormat}
      createMutationsPromises={createMutationsPromises(client, discussionPhaseId)}
      save={save}
      validate={validate}
      mutators={{
        ...arrayMutators,
        swap: swap
      }}
      render={({ handleSubmit, submitting, values, initialValues }) => {
        // the form is splitted to multiple subforms: We must calculate the global pristine.
        // Don't use final form pristine here
        const pristine = isEqualWith(initialValues, values, compareEditorState);
        return (
          <React.Fragment>
            <div className="admin-content">
              <AdminForm handleSubmit={handleSubmit} pristine={pristine} submitting={submitting}>
                {section === '1' && <Step1 editLocale={editLocale} locale={locale} discussionPhaseId={discussionPhaseId} />}
                {section === 'configThematics' && (
                  <ConfigureThematicForm
                    pristine={pristine}
                    phaseIdentifier={phaseIdentifier}
                    thematicId={thematicId}
                    editLocale={editLocale}
                    values={values}
                  />
                )}
              </AdminForm>
            </div>
          </React.Fragment>
        );
      }}
    />
  );
};

const mapStateToProps = state => ({
  debateId: state.context.debateId,
  editLocale: state.admin.editLocale,
  locale: state.i18n.locale
});

export default compose(
  connect(mapStateToProps),
  withApollo
)(DumbSurveyAdminForm);