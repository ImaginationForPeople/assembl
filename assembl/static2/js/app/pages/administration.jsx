import React from 'react';
import { compose, graphql } from 'react-apollo';
import { filter } from 'graphql-anywhere';
import { Grid, Row, Col } from 'react-bootstrap';
import { connect } from 'react-redux';

import { updateThematics } from '../actions/adminActions';
import withLoadingIndicator from '../components/common/withLoadingIndicator';
import Menu from '../components/administration/menu';
import LanguageMenu from '../components/administration/languageMenu';
import SaveButton from '../components/administration/saveButton';
import ThematicsQuery from '../graphql/ThematicsQuery.graphql';

class Administration extends React.Component {
  constructor() {
    super();
    this.putThematicsInStore.bind(this);
  }

  componentDidMount() {
    this.putThematicsInStore(this.props.data);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.data.thematics !== this.props.data.thematics) {
      this.putThematicsInStore(nextProps.data);
    }
  }

  putThematicsInStore(data) {
    // filter with the same query to remove stuff like __typename from the structure
    const filteredThematics = filter(ThematicsQuery, data);
    this.props.updateThematics(filteredThematics.thematics);
  }

  render() {
    const { children, data, params } = this.props;
    const { phase } = params;
    return (
      <div className="administration">
        <div className="save-bar">
          <div className="max-container">
            <Grid fluid>
              <Row>
                <Col xs={12} md={3} />
                <Col xs={12} md={8}>
                  <SaveButton refetchThematics={data.refetch} />
                </Col>
                <Col xs={12} md={1} />
              </Row>
            </Grid>
          </div>
        </div>
        <div className="max-container">
          <Grid fluid>
            <Row>
              <Col xs={12} md={3}>
                <div className="admin-menu-container">
                  <Menu requestedPhase={phase} />
                </div>
              </Col>
              <Col xs={12} md={8}>{children}</Col>
              <Col xs={12} md={1}>
                <LanguageMenu />
              </Col>
            </Row>
          </Grid>
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    updateThematics: (thematics) => {
      return dispatch(updateThematics(thematics));
    }
  };
};

export default compose(
  graphql(ThematicsQuery, {
    options: { variables: { identifier: 'survey' } }
  }),
  withLoadingIndicator,
  connect(null, mapDispatchToProps)
)(Administration);