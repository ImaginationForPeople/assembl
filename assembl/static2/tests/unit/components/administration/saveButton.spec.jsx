import React from 'react';
import renderer from 'react-test-renderer';

import * as saveButton from '../../../../js/app/components/administration/saveButton';
import '../../../helpers/setupTranslations';

describe('runSerial function', () => {
  const { runSerial } = saveButton;
  it('should run promises in serial', async () => {
    const task1 = jest.fn();
    const task2 = jest.fn();
    await runSerial([task1, task2]).then(() => {
      expect(task1.mock.calls.length).toBe(1);
      expect(task2.mock.calls.length).toBe(1);
    });
  });
});

describe('getMutationsPromises', () => {
  const { getMutationsPromises } = saveButton;
  it('should return an empty list if items is empty', () => {
    const params = {
      items: []
    };
    const result = getMutationsPromises(params);
    expect(result).toEqual([]);
  });

  it('should return a list with create mutations on new items', () => {
    const createMutationSpy = jest.fn();
    const deleteMutationSpy = jest.fn();
    const params = {
      createMutation: createMutationSpy,
      deleteMutation: deleteMutationSpy,
      items: [
        {
          _hasChanged: false,
          _isNew: true,
          _toDelete: false,
          title: 'Application'
        }
      ],
      variablesCreator: (item, idx) => ({
        title: item.title,
        idx: idx
      })
    };
    const result = getMutationsPromises(params);
    expect(result.length).toBe(1);
    result.forEach(task => task());
    expect(deleteMutationSpy.mock.calls.length).toBe(0);
    expect(createMutationSpy.mock.calls.length).toBe(1);
    expect(createMutationSpy.mock.calls[0][0]).toEqual({
      variables: {
        title: 'Application',
        idx: 0
      }
    });
  });

  it('should return a list with delete mutations on deleted items', () => {
    const createMutationSpy = jest.fn();
    const deleteMutationSpy = jest.fn();
    const params = {
      createMutation: createMutationSpy,
      deleteMutation: deleteMutationSpy,
      items: [
        {
          _hasChanged: false,
          _isNew: false,
          _toDelete: true,
          id: 1
        },
        {
          _hasChanged: false,
          _isNew: false,
          _toDelete: false,
          id: 2
        },
        {
          _hasChanged: false,
          _isNew: true,
          _toDelete: true,
          id: 3
        }
      ],
      deleteVariablesCreator: item => ({
        id: item.id
      })
    };
    const result = getMutationsPromises(params);
    expect(result.length).toBe(1);
    result.forEach(task => task());
    expect(createMutationSpy.mock.calls.length).toBe(0);
    expect(deleteMutationSpy.mock.calls.length).toBe(1);
    expect(deleteMutationSpy.mock.calls[0][0]).toEqual({
      variables: { id: 1 }
    });
  });

  it('should return a list with update mutations on items that have changed', () => {
    const updateMutationSpy = jest.fn();
    const params = {
      updateMutation: updateMutationSpy,
      items: [
        {
          _hasChanged: false,
          _isNew: false,
          _toDelete: false,
          id: 1
        },
        {
          _hasChanged: false,
          _isNew: true,
          _toDelete: false,
          id: 2
        },
        {
          _hasChanged: true,
          _isNew: false,
          _toDelete: false,
          id: 3
        },
        {
          _hasChanged: true,
          _isNew: false,
          _toDelete: false,
          id: 4
        }
      ],
      lang: 'en',
      deleteVariablesCreator: () => {},
      variablesCreator: item => ({ id: item.id })
    };
    const result = getMutationsPromises(params);
    expect(result.length).toBe(2);
    result.forEach(task => task());
    expect(updateMutationSpy.mock.calls.length).toBe(2);
    expect(updateMutationSpy.mock.calls[0][0]).toEqual({
      variables: { id: 3, lang: 'en' }
    });
    expect(updateMutationSpy.mock.calls[1][0]).toEqual({
      variables: { id: 4, lang: 'en' }
    });
  });

  it('should not create mutations for items that are new and to delete', () => {
    const createMutationSpy = jest.fn();
    const deleteMutationSpy = jest.fn();
    const updateMutationSpy = jest.fn();
    const params = {
      createMutation: createMutationSpy,
      deleteMutation: deleteMutationSpy,
      updateMutation: updateMutationSpy,
      items: [
        {
          _hasChanged: false,
          _isNew: true,
          _toDelete: true,
          id: 1
        },
        {
          _hasChanged: true,
          _isNew: true,
          _toDelete: true,
          id: 2
        }
      ],
      lang: 'en',
      deleteVariablesCreator: item => ({ id: item.id }),
      variablesCreator: item => ({ id: item.id })
    };
    const result = getMutationsPromises(params);
    expect(result.length).toBe(0);
    result.forEach(task => task());
    expect(createMutationSpy.mock.calls.length).toBe(0);
    expect(deleteMutationSpy.mock.calls.length).toBe(0);
    expect(updateMutationSpy.mock.calls.length).toBe(0);
  });
});

describe('Save button component', () => {
  const Component = saveButton.default;
  const saveActionSpy = jest.fn();
  it('should render a save button', () => {
    const props = {
      saveAction: saveActionSpy
    };
    const component = renderer.create(<Component {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a disabled save button', () => {
    const props = {
      disabled: true,
      saveAction: saveActionSpy
    };
    const component = renderer.create(<Component {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a save button with specific classes', () => {
    const props = {
      saveAction: saveActionSpy,
      specificClasses: 'my-great-green-save-button'
    };
    const component = renderer.create(<Component {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});