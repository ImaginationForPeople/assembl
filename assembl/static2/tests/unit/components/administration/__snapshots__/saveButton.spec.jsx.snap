// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Save button component should render a disabled save button 1`] = `
<button
  className="save-button button-submit button-dark right btn btn-default"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  <span
    className={undefined}
    style={undefined}
  >
    Save
  </span>
</button>
`;

exports[`Save button component should render a save button 1`] = `
<button
  className="save-button button-submit button-dark right btn btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span
    className={undefined}
    style={undefined}
  >
    Save
  </span>
</button>
`;

exports[`Save button component should render a save button with specific classes 1`] = `
<button
  className="my-great-green-save-button btn btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span
    className={undefined}
    style={undefined}
  >
    Save
  </span>
</button>
`;
